src/CombFilter/CombFilter.cpp

```diff
@@ -0,0 +1,138 @@
+#include "CombFilter.h"
+
+CCombFilterBase::CCombFilterBase(float fMaxDelayLengthInS, float fSampleRateInHz, int iNumChannels)
+: m_fGain(0.F), m_iNumChannels(iNumChannels), m_buffer(nullptr), m_fSampleRateInHz(fSampleRateInHz)
+{
+    setDelay(fMaxDelayLengthInS);
+}
+
+CCombFilterBase::~CCombFilterBase() {
+    clearRingBuffer();
+}
+
+bool CCombFilterBase::isInRange(float lower, float upper, float value) {
+    return (lower <= value && value < upper);
+}
+
+void CCombFilterBase::clearRingBuffer() {
+    if (m_buffer != nullptr) {
+        for (int i = 0; i < m_iNumChannels; ++i) {
+            if (m_buffer[i] != nullptr) {
+                delete m_buffer[i];
+                m_buffer[i] = nullptr;
+            }
+        }
+        delete[] m_buffer;
+    }
+    m_buffer = nullptr;
+}
+
+/*
+Setter part
+*/
+Error_t CCombFilterBase::setGain(float fGain) {
+    if (!isInRange(-m_kfMaxGain, m_kfMaxGain, fGain)) {
+        return Error_t::kFunctionInvalidArgsError;
+    }
+    m_fGain = fGain;
+    return Error_t::kNoError;
+}
+
+/*! Update the delay length of filter. This would cause deleting & newing CRingBuffer object. 
+\param fDelay the delay time IN SECOND
+*/
+Error_t CCombFilterBase::setDelay(float fDelay) {
+    if (!isInRange(0.F, m_kfMaxDelay, fDelay)) {
+        return Error_t::kFunctionInvalidArgsError;
+    }
+	m_fDelay = fDelay;
```

Comment:
Sorry but I don't quite understand this. Do you mean that the CRingBuffer length is 3, yet we only maintain 1 sample in it? It's feasible. You can maintain the number of samples by correctly call methods. I'm simply resetting the ring buffer, because if the new delay is longer than current one, then I have to enlarge the size of buffer, which is only possible by `new` another consecutive part of space. Therefore no matter how, I have to reallocate (that is `delete` & `new` ring buffer). If I don't reset, then I have to do things like copy & pasting content from one buffer to another, up-sampling & down-sampling, which is beyond this assignment.

Reply:
I meant that the CRingBuffer length is 3, it could be a nice addition to allow functionality where the playback start time of the entire 3 sample buffer ranges from e.g. 1-3